### ISSUE TITLE
Add spacing rules

### ISSUE THREAD

**Issue by maxmurdoch** [gocardless/javascript-style-guide#8]
It would be nice to get consistent when adding spacing around code blocks. For example, spaces here are probably unnecessary:

```js
function1(var1, var2) {

  function2('some_string', function(var3, var4) {
    …
  });

  return;
}
```

but spaces deeply indented are useful:

```js
]).config([
  '$stateProvider',
  function accessTokensCreateRoute($stateProvider) {
    $stateProvider.state(ACCESS_TOKENS_CREATE_ROUTE_STATE, {
      url: '/create',
      onEnter: [
        '$state',
        'Dialog',
        function($state, Dialog) {
   
         Dialog.show({
            template: template,
            controller: 'AccessTokensCreateController',
            controllerAs: 'ctrl',
            dialogHeader: 'Create Access Token',
            onDismissDialog: function onDismissDialog() {
              $state.go(ACCESS_TOKENS_INDEX_ROUTE_STATE);
            },
          });

        },
      ],
    });
  },
]);
```

In some places we’re adding a space above but not below a chunk of code. I think it should be both or neither. Any other spacing issues, feel free to add 'em.

Thoughts @jackfranklin @harrison @samjewell @jamesshedden?

**Comment by jackfranklin** [comment]
:+1: I don't really mind which way to do it. The difficulty enforcing this is that I find more than with anything else, the whitespace is down to context and hence one rule for all doesn't necessarily cover it.

Happy to go with the general consensus.

**Comment by jamesshedden** [comment]
Like Jack says I don't think one rule sounds feasible but it seems like we could come up with a set of rules/considerations. In my (limited) experience it seems to be strongly tied to how much is being nested, how many levels of nesting there are etc, would definitely be nice to pin down some guidelines if it makes sense.

**Comment by maxmurdoch** [comment]
I agree. It's tricky. My proposed rules would be:

1) Always have even spacing.

```js
// bad
function partialApplication(var1) {

  return var2 => var1 + var2;
}
// good
function partialApplication(var1) {
  return var2 => var1 + var2;
}
// or
function partialApplication(var1) {

  return function timeToApply(var2) {
    …
  }

}
```

2) Use space when nested further than 3 (?) levels deep. We could give this a go and then adjust dependant on how it feels?

**Comment by harrison** [comment]
1. :+1: :+1: 
2. Not obvious what the right way is, do you only ever space out the most deeply nested func/block?

Personally not too concerned with 2 as long as we follow 1:)

**Comment by jamesshedden** [comment]
Some rules I follow with HTML which work pretty well to do with nesting is like this:

Only one block nested, and only one-liners:

eg
```
<div>
  <div></div>
</div>
```

```
<div>
  <div></div>
  <div></div>
  <div></div>
</div>
```

This can work infinitely in the case of one nested block that simply contains another nested block:

```
<div>
  <div>
    <div>
      <div></div>
    </div>
  <div>
</div>
```

At a certain level of complexity, the above might begin to be a bit unreadable, which is where it starts to come down to making a call depending on context, but in practice it's not too bad and it's mainly when dealing with multiple child blocks that things become hard to parse, leading me on to...

When dealing with multiple nested blocks in any situation, the top and bottom-most lines don't need spacing to separate them from their parent block, but each child block should have spaces between it:

```
<div>
  <div>
    <div></div>
  </div>

  <div>
    <div></div>
  </div>
</div>
```

This all might be really obvious but it's something close to a set of rules that might also work quite nicely with JavaScript?

**Comment by samjewell** [comment]
This sounds great to me, awesome. We can always tweak it, but would be great to get this in. Can we put in the HTML stuff also somewhere - maybe kick off that HTML styleguide?
